text/0000-try-trait-v2.md

```diff
@@ -0,0 +1,999 @@
+        match cf {
+            ControlFlow::Continue(a) => accum = a,
+            ControlFlow::Break(_) => todo!(),
+        }
+    }
+    R::continue_with(accum)
+}
+```
+
+Just `Bubble` isn't enough, though.  We also need `Try` here in order to convert the holder inside the `Break` variant back into `R`.  But because `Bubble` is a supertrait of `Try<H>`, we can just change the bound rather than adding a second one:
+```rust
+pub fn simple_try_fold<A, T, R: Try<Continue = A>>(
+    iter: impl Iterator<Item = T>,
+    mut accum: A,
+    mut f: impl FnMut(A, T) -> R,
+) -> R {
+    for x in iter {
+        let cf = f(accum, x).branch();
+        match cf {
+            ControlFlow::Continue(a) => accum = a,
+            ControlFlow::Break(h) => return R::from_holder(h),
+        }
+    }
+    R::continue_with(accum)
+}
+```
+
+### When you need to switch to a different `Continue` type
+
+You might have noticed that even the final `simple_try_fold` didn't need to mention `BreakHolder<C>`.  That's because it's only needed if you want to return a type from the same family, but with a different continue type.
+
+`Option::zip` stabilized a few releases ago, so here's a demonstration of a more-generalized version.  To restrict the two inputs to the same family, it requires that the holder types match. Then it uses `BreakHolder` to get an output type from that same family.
+
+```rust
+fn zip_demo<T, U, H: BreakHolder<(T, U)>>(
+    a: impl Bubble<Continue = T, Holder = H>,
+    b: impl Bubble<Continue = U, Holder = H>,
+) -> <H as BreakHolder<(T, U)>>::Output {
+    Bubble::continue_with((a?, b?))
+}
+```
+
+Here `Try` doesn't need to be mentioned explicitly, as the `BreakHolder::Output` type is always bound by `Try<Self>`.
+
+<!--
+Explain the proposal as if it was already included in the language and you were teaching it to another Rust programmer. That generally means:
+
+- Introducing new named concepts.
+- Explaining the feature largely in terms of examples.
+- Explaining how Rust programmers should *think* about the feature, and how it should impact the way they use Rust. It should explain the impact as concretely as possible.
+- If applicable, provide sample error messages, deprecation warnings, or migration guidance.
+- If applicable, describe the differences between teaching this to existing Rust programmers and new Rust programmers.
+
+For implementation-oriented RFCs (e.g. for compiler internals), this section should focus on how compiler contributors should think about the change, and give examples of its concrete impact. For policy RFCs, this section should provide an example-driven introduction to the policy, and explain its impact in concrete terms.
+-->
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## `ops::ControlFlow`
+
+```rust
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum ControlFlow<B, C = ()> {
+    /// Exit the loop, yielding the given value
```

Comment:
Earlier I mention the use of "flow" as a more apt term and it may be helpful here. Flows can loop but don't have to.

Reply:
These comments were taken from nightly.  I've since PRed better ones, as well as more examples in the docs: https://doc.rust-lang.org/nightly/std/ops/enum.ControlFlow.html